x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/rules_table/use_rules_table.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Dispatch, useMemo, useReducer, useEffect, useRef } from 'react';
+import { EuiBasicTable } from '@elastic/eui';
+
+import { errorToToaster, useStateToaster } from '../../../../../common/components/toasters';
+import * as i18n from '../translations';
+
+import { fetchRules } from '../api';
+import { createRulesTableReducer, RulesTableState, RulesTableAction } from './rules_table_reducer';
+import { createRulesTableFacade, RulesTableFacade } from './rules_table_facade';
+import { useFilteredRules } from './projections/filtering';
+import { useSortedRules } from './projections/sorting';
+import { usePaginatedRules, useDisplayedPaginationOptions } from './projections/pagination';
+
+// Why these values?
+// If we request more rules than 10000, we'll get an error from Elasticsearch (most likely).
+//   https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html
+//   https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings
+//   see `index.max_result_window`, default value is 10000.
+// Example:
+//   In requests like `GET my-index/_search?from=40&size=20` it's not about items per page,
+//   but overall from+size count (requested page + all previous pages).
+//   If max_result_window=10000 this will fail: `GET my-index/_search?from=9990&size=20`.
+// We load "all" rules in memory at once. So, because we always request page=1:
+const MAX_RULES_TO_LOAD = 10000; // must not be increased
+const NUM_RULES_TO_LOAD = MAX_RULES_TO_LOAD / 2; // must be < MAX_RULES_TO_LOAD
```

Comment:
5000 is an arguable value, please make suggestions if you will.

Unfortunately, no specific value can give us full confidence that all the rules will be loaded in all cases for all Stack deployments and customers. For the time being, this should work, but we definitely need to switch back to server-side filtering later.

I should also note that loading our current 461 rules takes quite some time, and I'm not sure if loading 5000 real rules would be practical prom the perf/UX standpoint.